homeassistant/components/evohome.py

```diff
@@ -0,0 +1,606 @@
+    """
+    # Used for internal data, such as installation, state & timers...
+    hass.data[DATA_EVOHOME] = {}
+
+    domain_data = hass.data[DATA_EVOHOME]
+    domain_data['timers'] = {}
+
+    # Pull the configuration parameters...
+    domain_data['params'] = dict(config[DOMAIN])
+    # scan_interval - start with default value
+    domain_data['params'][CONF_SCAN_INTERVAL] = SCAN_INTERVAL_DEFAULT
+
+    if _LOGGER.isEnabledFor(logging.DEBUG):  # then redact username, password
+        tmp = dict(domain_data['params'])
+        tmp[CONF_USERNAME] = 'REDACTED'
+        tmp[CONF_PASSWORD] = 'REDACTED'
+
+        _LOGGER.debug("setup(): Configuration parameters: %s", tmp)
+
+    from evohomeclient2 import EvohomeClient
+
+    _LOGGER.debug("setup(): API call [4 request(s)]: client.__init__()...")
+
+    try:
+        # there's a bug in evohomeclient2 v0.2.7: the client.__init__() sets
+        # the root loglevel (debug=?), so must remember it now...
+        log_level = logging.getLogger().getEffectiveLevel()
+
+        client = EvohomeClient(
+            domain_data['params'][CONF_USERNAME],
+            domain_data['params'][CONF_PASSWORD],
+            debug=False
+        )
+        # ...then restore it to what it was before instantiating the client
+        logging.getLogger().setLevel(log_level)
+
+    except requests.RequestException as err:
+        if str(HTTP_BAD_REQUEST) in str(err):
+            # this happens when bad user credentials are supplied
+            _LOGGER.error(
+                "Failed to establish a connection with evohome web servers, "
+                "Check your username (%s), and password are correct."
+                "Unable to continue. Resolve any errors and restart HA.",
+                domain_data['params'][CONF_USERNAME]
+            )
+        else:
+            # Otherwise, it may be enough to back off and try again later.
+            raise PlatformNotReady(err)
+
+    finally:  # Redact username, password as no longer needed.
+        domain_data['params'][CONF_USERNAME] = 'REDACTED'
+        domain_data['params'][CONF_PASSWORD] = 'REDACTED'
+
+    domain_data['client'] = client
+
+    # Redact any installation data we'll never need.
+    if client.installation_info[0]['locationInfo']['locationId'] != 'REDACTED':
+        for loc in client.installation_info:
+            loc['locationInfo']['locationId'] = 'REDACTED'
+            loc['locationInfo']['streetAddress'] = 'REDACTED'
+            loc['locationInfo']['city'] = 'REDACTED'
+            loc['locationInfo']['locationOwner'] = 'REDACTED'
+            loc[GWS][0]['gatewayInfo'] = 'REDACTED'
+
+    # Pull down the installation configuration...
+    loc_idx = domain_data['params'][CONF_LOCATION_IDX]
+
+    try:
+        domain_data['config'] = client.installation_info[loc_idx]
+
+    # IndexError: configured location index is outside the range
+    except IndexError:
+        _LOGGER.warning(
```

Comment:
We should not recover and do something else. Instead we should fail setup. This is clearly a misconfiguration. 

Imagine removing a site from your account and all of a sudden all entities are replaced with something else. As a user you will expect it to fail.

Reply:
Fair enough.

Reply:
OK to do this?
```python
    except IndexError:
        _LOGGER.warning(
            "setup(): Config parameter, '%s'= %s , is out of range (0-%s)",
            CONF_LOCATION_IDX,
            loc_idx,
            len(client.installation_info) - 1
        )

        raise
```

Reply:
OK, leaving the `warning()` in, as the Traceback is pretty indecipherable without it.

Reply:
in the end, this was taken out